crates/oxc_minifier/src/compressor/mod.rs

```diff
@@ -0,0 +1,214 @@
+
+    /// Drop `drop_debugger` statement.
+    /// Enabled by `compress.drop_debugger`
+    fn drop_debugger<'b>(&mut self, stmt: &'b Statement<'a>) -> bool {
+        matches!(stmt, Statement::DebuggerStatement(_)) && self.options.drop_debugger
+    }
+
+    /// Join consecutive var statements
+    fn join_vars<'b>(&mut self, stmts: &'b mut Vec<'a, Statement<'a>>) {
+        // Collect all the consecutive ranges that contain joinable vars.
+        // This is required because Rust prevents in-place vec mutation.
+        let mut ranges = vec![];
+        let mut range = 0..0;
+        let mut i = 1usize;
+        let mut capacity = 0usize;
+        for window in stmts.windows(2) {
+            let [prev, cur] = window else { unreachable!() };
+            if let Statement::Declaration(Declaration::VariableDeclaration(cur_decl)) = cur
+                && let Statement::Declaration(Declaration::VariableDeclaration(prev_decl)) = prev
+                && cur_decl.kind == prev_decl.kind {
+                if i - 1 != range.end  {
+                    range.start = i - 1;
+                }
+                range.end = i + 1;
+            }
+            if (range.end != i || i == stmts.len() - 1) && range.start < range.end {
+                capacity += range.end - range.start - 1;
+                ranges.push(range.clone());
+                range = 0..0;
+            }
+            i += 1;
+        }
+
+        if ranges.is_empty() {
+            return;
+        }
+
+        // Reconstruct the stmts array by joining consecutive ranges
+        let mut new_stmts = self.ast.new_vec_with_capacity(stmts.len() - capacity);
+        for (i, stmt) in stmts.drain(..).enumerate() {
+            if i > 0
+                && ranges.iter().any(|range| range.contains(&i))
+                && let Statement::Declaration(Declaration::VariableDeclaration(prev_decl)) = new_stmts.last_mut().unwrap()
+                && let Statement::Declaration(Declaration::VariableDeclaration(mut cur_decl)) = stmt {
+                prev_decl.declarations.append(&mut cur_decl.declarations);
+            } else {
+                new_stmts.push(stmt);
+            }
+        }
+        *stmts = new_stmts;
+    }
+
+    /// Transforms `while(expr)` to `for(;expr;)`
+    fn compress_while<'b>(&mut self, stmt: &'b mut Statement<'a>) {
+        if let Statement::WhileStatement(while_stmt) = stmt
+            && self.options.loops {
+            let dummy_test = self.ast.this_expression(SPAN);
+            let test = std::mem::replace(&mut while_stmt.test, dummy_test);
+            let dummy_body = self.ast.empty_statement(SPAN);
+            let body = std::mem::replace(&mut while_stmt.body, dummy_body);
+            *stmt = self.ast.for_statement(SPAN, None, Some(test), None, body);
+        }
+    }
+
+    /* Expressions */
+
+    /// Transforms `undefined` => `void 0`
+    fn compress_undefined<'b>(&mut self, expr: &'b mut Expression<'a>) -> bool {
+        if expr.is_undefined() {
```

Comment:
I think you need to check if it is shadowed.

Reply:
No worries, this is just a prototype.